File: sdk/cpp/examples/vision/resize.cpp
// OA_DOC_BEGIN: vision-resize
#include <oa/core/fnMatrix.h>
#include <oa/core/image.h>
#include <oa/runtime/engine.h>
#include <oa/vision/fnImage.h>

#include <array>
#include <cmath>
#include <cstdio>
#include <utility>

int main() {
	oa::EngineConfig config;
	config.appName = "ExampleVisionResize";
	config.presentationMode = oa::PresentationMode::None;

	auto created = oa::Engine::create(config);
	if (not created.isOk()) return 1;
	auto engine = std::move(created).getValue();

	oa::Image image(
		oa::FnMatrix::full({1, 3, 2, 2}, 0.25F),
		oa::ImageLayout::Nchw,
		oa::ImageFormat::Rgb);
	auto resized = oa::FnImage::resize(image, 4, 3);

	auto submitted = engine->submit();
	if (not submitted.isOk()) return 1;
	if (not engine->wait(submitted.getValue()).isOk()) return 1;
	if (not resized.validate() || resized.width() != 4 || resized.height() != 3) {
		return 1;
	}

	std::array<oa::F32, 36> values{};
	if (not oa::FnMatrix::copyToHost(
		resized.asMatrix(), values.data(), sizeof(values)).isOk()) return 1;
	for (const oa::F32 value : values) {
		if (std::abs(value - 0.25F) > 1.0e-6F) return 1;
	}

	std::puts("RGB NCHW image resized from 2x2 to 4x3");
	return 0;
}
// OA_DOC_END: vision-resize

File: sdk/py/examples/vision/resize.py
#!/usr/bin/env python3
"""Resize a semantic image on the GPU and preserve its metadata.
"""

# OA_DOC_BEGIN: vision-resize
from oa import (
	FnImage,
	FnMatrix,
	Image,
	ImageFormat,
	ImageLayout,
	initComputeEngine,
	shutdownComputeEngine,
)


if not initComputeEngine():
	raise RuntimeError("OA could not create a Vulkan compute engine")

try:
	image = Image(
		FnMatrix.full([1, 3, 2, 2], 0.25),
		ImageLayout.Nchw,
		ImageFormat.Rgb,
	)
	resized = FnImage.resize(image, 4, 3)
	values = FnMatrix.copyToHost(resized.asMatrix())
	assert resized.validate()
	assert resized.width() == 4
	assert resized.height() == 3
	assert values == [0.25] * 36
	print("RGB NCHW image resized from 2x2 to 4x3")
finally:
	shutdownComputeEngine()
# OA_DOC_END: vision-resize
